<h1 id="crisprroots">CRISPRroots</h1>
<p>CRISPRroots: <strong>R</strong>NA-seq based <strong>o</strong>n-target and <strong>o</strong>ff-<strong>t</strong>arget assessment 
of Cas9-mediated edit<strong>s</strong></p>
<p>The CRISPR/Cas9 genome editing tool can be used to study genomic variants and gene knockouts.
By combining CRISPR/Cas9 mediated editing with transcriptomic analyses it is possible to measure
the effects of genome alterations on gene expression. In such experiments it is crucial to understand
not only if the editing was successful but also if the observed differential gene expression is the result
of intended genome edits and not that of unwanted off-target effects. Potential off-targets sites that
CRISPR/Cas9 may have edited need therefore to be verified by sequencing. However, a CRISPR/Cas9
gRNA may have hundreds (or thousands) potential off-target binding sites in a genome. How to decide
which of them should be validated with higher priority? The RNA-seq data that can be sequenced
as part of a CRISPR/Cas9 experiment contains information about the sequence and expression level
of potential off-target sites/genes located in transcribed regions. Here we preset CRISPRroots, a
method that combines CRISPR/Cas9 and guide RNA binding properties, gene expression changes,
and sequence variants between edited and non-edited cells, to discover and rank potential off-targets.
The method is described in the corresponding publication (see below).</p>
<h2 id="installation-and-configuration">Installation and configuration</h2>
<h3 id="prerequisites">Prerequisites</h3>
<p>The CRISPRroots pipeline can be executed via 
<a href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a>.<br>To be executed in Sankemake, the pipeline requires:</p>
<ul>
<li>conda : ≥4.8.5</li>
<li>snakemake : ≥5.32.00 </li>
</ul>
<p>All other software requirements are satisfied by the Conda environments defined in Snakemake,
which are installed by starting Snakemake with the <code>--use-conda</code> flag.</p>
<p>The pipeline was tested in a x86 64 GNU/Linux environment with Ubuntu v.18.04.1 (or newer) 
installed.</p>
<h3 id="setup">Setup</h3>
<p>CRISPRroots is available to download from 
<a href="https://rth.dk/resources/crispr/">https://rth.dk/resources/crispr/</a>.<br>A test dataset is also available on the same website. 
After downloading and un-packing the software and the test dataset, we recommend 
to explore the directory structure of the test dataset.</p>
<ul>
<li>CRISPRroots_test_dataset<ul>
<li>resources #reference files</li>
<li>QPRT_DEL268T_chr16_10M-40M #sample directory</li>
<li><code>make_config.py</code> #setup config file</li>
</ul>
</li>
</ul>
<p>the script <code>make_config.py</code> can be used to automatically set the paths to the data, resources, 
and code directories in the configuration file <code>config.yaml</code> located in 
<code>CRISPRroots_test_dataset/QPRT_DEL268T_chr16_10M-40M</code>. 
To run the script, you need python3.
To setup the config file for the test dataset, run:</p>
<pre><code class="lang-shell">  <span class="hljs-keyword">cd</span> CRISPRroots_test_dataset
  <span class="hljs-keyword">python3</span> make_config.<span class="hljs-keyword">py</span> --CRISPRroots <span class="hljs-symbol">&lt;path_to_CRISPRroots&gt;</span>
</code></pre>
<p>The config file contains the parameters defined for the execution of the various steps of the pipeline. 
A copy of the config file for the CRISPRroots test dataset is provided together with 
the CRISPRroots software package. This file can be used as template to create the configuration file 
for your own dataset.</p>
<p>For a complete list of parameters, options, and usage examples for CRISPRroots please read the 
<em>CRISPRroots_Manual.pdf</em> included in the CRISPRroots software folder.</p>
<h2 id="pipeline-usage">Pipeline usage</h2>
<h3 id="basic-usage">Basic Usage</h3>
<p>Assuming you have installed the software in the prerequisites, you can run the pipeline 
from within the directory containing the config file (in the test dataset this is
the subfolder <em>QPRT_DEL268T_chr16_10M-40M</em>) as follows: </p>
<pre><code class="lang-shell">  <span class="hljs-keyword">cd</span> QPRT_DEL268T_chr16_10M-40M #<span class="hljs-keyword">for</span> usage <span class="hljs-keyword">in</span> the <span class="hljs-keyword">test</span> dataset, substitute with your own directory otherwise
  snakemake -s &lt;path_to_CRISPRroots&gt;/<span class="hljs-keyword">run</span>.smk --<span class="hljs-keyword">use</span>-conda --dry-<span class="hljs-keyword">run</span> 
  snakemake -s &lt;path_to_CRISPRroots&gt;/<span class="hljs-keyword">run</span>.smk --<span class="hljs-keyword">use</span>-conda --cores &lt;int&gt;
</code></pre>
<p>We recommend to first run Snakemake with <code>--dryrun</code>.<br>This displays what will be done without executing it and highlights if any input file is missing. 
In the commands above, <code>--cores</code> specifies the maximum number of cores used in parallel by Snakemake.
The pre-computed results/reports for the test dataset are available in the folder 
<em>CRISPRroots_test_dataset/QPRT_DEL268T_chr16_10M-40M/pre-computed</em>.</p>
<p>Hint: Snakemake allows to visualize the jobs as a graph (directed acyclic graph, or DAG), 
highlighting the jobs completed and those to be run in different ways. 
To create an <em>svg</em> plot of your DAG, run the following command:</p>
<pre><code class="lang-shell">    snakemake -s &lt;path_to_CRISPRroots&gt;/<span class="hljs-built_in">run</span>.smk <span class="hljs-comment">--dag | dot -Tsvg &gt; dag.svg</span>
</code></pre>
<p>To learn more about the DAG and the visualization of jobs please visit the Snakemake tutorial
at <a href="https://snakemake.readthedocs.io/en/stable/tutorial/basics.html">https://snakemake.readthedocs.io/en/stable/tutorial/basics.html</a> (<em>Step 4: Indexing read alignments and visualizing the DAG of jobs</em>).</p>
<h3 id="advanced-usage">Advanced usage</h3>
<p>The pipeline can also be used to accomplish only specific tasks. 
For example, to only perform the the pre-processing of the dataset and read mapping, 
you can run CRISPRroots with the rule flag <code>preproc_and_map</code> at the end:</p>
<pre><code class="lang-shell">  snakemake -s &lt;path_to_CRISPRroots/<span class="hljs-built_in">run</span>.smk <span class="hljs-comment">--use-conda --cores &lt;int&gt; preproc_and_map</span>
</code></pre>
<p>The tasks (Snakemake rules) ready for use are the following:</p>
<ul>
<li><strong>variants_to_genome</strong>: executes the rules necessary to produce files containing 
filtered variants between each sample and the reference genome. 
Output in: <code>&lt;path_to_results_folder&gt;/6_GATK_variants/&lt;sample name&gt;/variants_filtered.vcf</code></li>
<li><strong>eSNPKaryotyping</strong>: Executes the R package eSNP-Karyotyping for the analysis of 
genome integrity from RNA-seq. 
The standard workflow is modified to employ reads mapped with STAR instead of TopHat2. 
Output in: <code>&lt;path_to_report_folder&gt;/eSNPKaryotyping/</code></li>
<li><strong>on_target_check</strong>: executes the on-target editing assessment. 
Output in: <code>&lt;path_to_report_folder&gt;/on_target_knockin.xlsx</code>; 
<code>&lt;path_to_report_folder&gt;/on_target_knockout.xlsx</code></li>
<li><strong>get_variated_genome</strong>: produces a variant-aware version of the reference genome, 
in which variants discovered from the RNA-seq are introduced in the reference sequence. 
Output in: <code>&lt;path_to_results_folder&gt;/6_GATK_variants/variated_genome.fa</code></li>
<li><strong>get_lib_type</strong>: assesses the library type with RSeQC. 
Output in: <code>&lt;path_to_results_folder&gt;/2-1_RSeQC_libtype/</code></li>
<li><strong>preproc_and_map</strong>: runs the reads pre-processing and mapping.<br>Mapping output in: <code>&lt;path_to_results_folder&gt;/2_sortaligned/</code><br>Mapping statistics in: <code>&lt;path_to_report_folder&gt;/report/mapping_stats.xlsx</code><br>Pre-processing results in: <code>&lt;path_to_results_folder&gt;/preproc/</code><br>Pre-processing statistics in: <code>&lt;path_to_report_folder&gt;/report/multiqc_samples_stats.xlsx</code> </li>
</ul>
<p>NB: In the test dataset, <code>&lt;path_to_results_folder&gt;</code> and <code>&lt;path_to_report_folder&gt;</code> correspond to the 
subfolders <em>results</em> and <em>report</em> in that will be generated inside 
<em>CRISPRroots_test_dataset/QPRT_DEL268T_chr16_10M-40M/</em> after completing the pipeline.</p>
<p>A useful flag for the execution of specific tasks is <code>--notemp</code>. 
This avoids removing output files defined as temporary in the pipeline 
(e.g. partially processed reads). It can be convenient to use it when executing only a 
part of the pipeline, to avoid the removal of temporary files that will need to be recreated 
if required by a subsequent execution of the pipeline.</p>
<h3 id="running-the-pipeline-in-a-computer-cluster">Running the pipeline in a computer cluster</h3>
<p>CRISPRroots can also be launched on a computer cluster. 
An example of how to set up CRISPRroots to run it with the Slurm Workload Manager 
is given in the test directory and can be used as:</p>
<pre><code class="lang-shell">  <span class="hljs-keyword">cd</span> CRISPRroots_test_dataset/QPRT_DEL268T_chr16_10M-<span class="hljs-number">40</span>M
  ./<span class="hljs-symbol">&lt;path_to_CRISPRroots&gt;</span>/cluster_run.<span class="hljs-keyword">sh</span>
</code></pre>
<p>You can add the name of a target rule to run only a part of the pipeline as below:</p>
<pre><code class="lang-shell">  ./&lt;path_to_CRISPRroots&gt;/cluster_run<span class="hljs-selector-class">.sh</span> [target_rule]
</code></pre>
<h2 id="output-files">Output files</h2>
<p>The pipeline’s output files are collected in two folders: <strong>report</strong> and <strong>results</strong>. 
Additionally, Snakemake generates a hidden folder, <strong>.snakemake</strong>, at the moment it is executed. 
Here, Snakemake stores all the information necessary to track the activity of the pipeline and 
the origin of each file it generates. The conda environments created by Snakemake are also stored in this folder.</p>
<p>The <strong>report</strong> folder contains the main output, including the candidate off-targets and the knockin/knockout
assessment. Results regarding differential expression and processing statistics (data quality and map-
ping) are also present.</p>
<p>Please refer to the <em>CRISPRroots_Manual.pdf</em> included in the CRISPRroots software folder for a complete description fo the 
output and its content.</p>
<h2 id="copyright">Copyright</h2>
<p>Copyright 2021 by the contributors:</p>
<p>Giulia Corsi <a href="&#x6d;&#97;&#x69;&#x6c;&#x74;&#111;&#x3a;&#x67;&#105;&#117;&#108;&#x69;&#x61;&#x40;&#114;&#x74;&#104;&#46;&#x64;&#107;">&#x67;&#105;&#117;&#108;&#x69;&#x61;&#x40;&#114;&#x74;&#104;&#46;&#x64;&#107;</a>, Veerendra Gadekar <a href="&#109;&#x61;&#x69;&#108;&#116;&#111;&#x3a;&#x76;&#101;&#101;&#114;&#64;&#114;&#x74;&#104;&#46;&#100;&#107;">&#x76;&#101;&#101;&#114;&#64;&#114;&#x74;&#104;&#46;&#100;&#107;</a></p>
<p>GNU GENERAL PUBLIC LICENSE</p>
<p>This is a free software: you can redistribute it and/or modify it under the terms of the GNU 
General Public License, either version 3 of the License, or (at your option) any later version. 
See <a href="https://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a>.</p>
<p>This software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.</p>
<h2 id="citations">Citations</h2>
<p>If you use CRISPRroots in your publication please cite:<br><strong>CRISPRroots: on- and off-target assessment of RNA-seq data in CRISPR-Cas9 edited cells</strong><br>Corsi GI, Gadekar VP, Gorodkin J, Seemann SE. <em>Submitted</em>.</p>
<h2 id="contact">Contact</h2>
<p>In case of problems or bug reports, please contact <a href="&#x6d;&#97;&#105;&#108;&#116;&#x6f;&#58;&#x73;&#x6f;&#x66;&#x74;&#x77;&#97;&#x72;&#101;&#x2b;&#x63;&#114;&#x69;&#115;&#x70;&#114;&#x72;&#111;&#x6f;&#116;&#115;&#x40;&#114;&#116;&#x68;&#46;&#100;&#107;">&#x73;&#x6f;&#x66;&#x74;&#x77;&#97;&#x72;&#101;&#x2b;&#x63;&#114;&#x69;&#115;&#x70;&#114;&#x72;&#111;&#x6f;&#116;&#115;&#x40;&#114;&#116;&#x68;&#46;&#100;&#107;</a></p>

